### PR TITLE
USB passthrough for VMs

### PR DESCRIPTION
**Pull request by monstermunchkin**

### CODE REVIEW

**Comment by stgraber** [comment]
```
[    2.521152] usb 5-1: new high-speed USB device number 2 using xhci_hcd
[    2.679465] usb 5-1: New USB device found, idVendor=04f2, idProduct=b5ce, bcdDevice= 0.06
[    2.679467] usb 5-1: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[    2.679468] usb 5-1: Product: Integrated Camera
[    2.679468] usb 5-1: Manufacturer: Chicony Electronics Co.,Ltd.
[    2.679804] usb 5-1: can't set config #1, error -32
```

**Comment by stgraber** [comment]
```
/:  Bus 01.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/12p, 480M
    ID 1d6b:0002 Linux Foundation 2.0 root hub
    |__ Port 8: Dev 2, If 0, Class=Video, Driver=, 480M
        ID 04f2:b5ce Chicony Electronics Co., Ltd Integrated Camera
    |__ Port 8: Dev 2, If 1, Class=Video, Driver=, 480M
        ID 04f2:b5ce Chicony Electronics Co., Ltd Integrated Camera
```

```
# USB host device ("webcam" device)
[device "dev-lxd_webcam"]
driver = "usb-host"
bus = "qemu_usb.0"
hostaddr = "2"
hostbus = "1"
```

**Comment by stgraber** [comment]
I wonder if in this case, hostaddr=1 hostbus=1 is what we should be using maybe?

**Comment by stgraber** [comment]
I did some manual testing and no, the device selection done in this branch is correct but for some reason we're getting:
```
[    1.441672] usb 5-1: can't set config #1, error -32
```

**Comment by stgraber** [comment]
`usb = on` isn't needed, it just causes a bunch more buses to show up than we need.

**Comment by stgraber** [comment]
@tomponline can you sync with @monstermunchkin on this? Maybe you've got some idea I haven't had yet :)

In short, we need to ensure that `lspci` only lists a single XHCI controller, dropping the `usb = on ` should take care of that.
Then we need to figure out why configuration isn't possible as that's preventing the devices from being usable...

So far I've tested running as real root and disabling sandboxing, neither helped.

**Comment by tomponline** [comment]
@stgraber @monstermunchkin yep will check this branch out and see if I can come up with any suggestions.

**Comment by monstermunchkin** [comment]
The problem is caused by the `-chroot` flag. When removing it, I can happily talk with the USB device.

**Comment by stgraber** [comment]
Hmm, any idea how to avoid that problem?

**Comment by tomponline** [comment]
@stgraber im trying to copy the usb unix dev from /dev/sys/bus into the virtual machine directory and potentially chown it so it works with -runas flag too.

**Comment by monstermunchkin** [comment]
> Hmm, any idea how to avoid that problem?

Looking into bind-mounting some relevant paths and/or files.

**Comment by tomponline** [comment]
The `-32` error means broken pipe apparently.

**Comment by stgraber** [comment]
They don't support us passing an fd through for that one?

**Comment by tomponline** [comment]
@stgraber I did think about that, but apparently not, also looking at the fds opened by qemu, it has the /dev/usb/bus device opened after start anyway, so looks like it switches chroot after opening it. Have also tried removing `runas` as @monstermunchkin has had some luck with getting the vendorid/productid mode working running as root, but this makes no different when using hostbus/hostaddr options.

**Comment by stgraber** [comment]
The full list of options on usb-host are:

 - hostbus
 - hostaddr
 - hostport
 - vendorid
 - productid
 - hostdevice (requires modern LIBUSB apparently)
 - isobufs (default 4)
 - isobsize (default 32)
 - guest-reset (default true)
 - guest-resets-all (default false)
 - loglevel (default warning)
 - pipeline (default true)
 - suppress-remote-wake (default true)

`hostdevice` sounds maybe promising?

File: lxd/apparmor/instance_qemu.go [hunk 1]
```diff
@@ -27,9 +27,13 @@ profile "{{ .name }}" flags=(attach_disconnected,mediate_deleted) {
...
   /etc/ceph/**                              r,
+  /run/udev/data/*                          r,
+  /sys/bus/                                 r,
   /sys/bus/nd/devices/                      r,
-  /sys/devices/system/node/                 r,
-  /sys/devices/system/node/**               r,
+  /sys/bus/usb/devices/                     r,
+  /sys/bus/usb/devices/**                   r,
+  /sys/class/                               r,
+  /sys/devices/**                           r,
```
Comment (stgraber) [inline lxd/apparmor/instance_qemu.go:36]: Alphabetical ordering would be nice.

**Pull request merged**